config.go

```diff
@@ -7,9 +7,89 @@ import (
 	"time"
 )
 
+// These are the versions of the protocol (which includes RPC messages as
+// well as Raft-specific log entries) that this server can _understand_. Use
+// the ProtocolVersion member of the Config object to control the version of
+// the protocol to use when _speaking_ to other servers. This is not currently
+// written into snapshots so they are unversioned. Note that depending on the
```

Comment:
This should probably be done as a separate PR, but we could add a snapshot version as well as a function similar to `RecoverCluster()` that doesn't take a new configuration but rolls up the logs and takes a snapshot, writing that out in the latest version. That tool should allow for us to deprecate things by saying "you must run this on all your servers prior to running X (it could maybe be in `NewRaft()` as something we do automatically for folks).